hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/protocol/datatransfer/sasl/SaslDataTransferClient.java

```diff
@@ -603,7 +603,17 @@ private IOStreamPair doSaslHandshake(InetAddress addr,
           conf, cipherOption, underlyingOut, underlyingIn, false) :
           sasl.createStreamPair(out, in);
     } catch (IOException ioe) {
-      sendGenericSaslErrorMessage(out, ioe.getMessage());
+      try {
```

Comment:
I found another case that slow might happen.

- Server failed to retrieve password and send accessTokenError to client
- Client receive ERROR + accessTokenError=true and InvalidBlockTokenException is thrown
- Reach to this catch clausure
- Client send generic sasl error message to server
- However server closes peer if exception happens in sasl handshake
  - https://github.com/apache/hadoop/blob/c9d64bad37fbfbaa77481f67338b39a029d7dc51/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/datanode/DataXceiver.java#L297-L336
- Client got IOException due to "Broken pipe"  and this IOException is thrown to caller instead of ioe which is InvalidToken
- So token is not refreshed and all datanodes are considered as dead nodes.

This is difficult to reproduce with current code because of the timing issue, but you can reproduce by changing this code to

```
    } catch (IOException ioe) {
      try {
        Thread.sleep(1000);
      } catch (InterruptedException e) {
        e.printStackTrace();
      }
      sendGenericSaslErrorMessage(out, ioe.getMessage());
      throw ioe;
    }
```

Then TestSaslDataTransferExpiredBlockToken will be failed.

We need to suppress exception from sendGenericSaslErrorMessage and throw ioe always

Reply:
Thank you for your information and additional fix! Can we create a flag to add `Thread.sleep` and add a unit test with the flag enabled?

Reply:
Thanks for the comment. somehow I couldn't reproduce the issue just sleeping and I dig into.
I could reproduce this issue when the message contains `accessBlockToken` always somehow finally.

Could you reproduce this issue or is this my local only?
Do you have any idea about this issue?
Commit to reproduce https://github.com/bitterfox/hadoop/commit/76fa47d545fc2b287a378431ddf482bfb467c4d6

Stacktrace:
```
2021-11-26 17:29:26,044 [hedgedRead-1] WARN  hdfs.DFSClient (DFSInputStream.java:actualGetFromOneDataNode(1257)) - Connection failure: Failed to connect to /127.0.0.1:45265 for file /file1 for block BP-1024827052-127.0.1.1-1637915363332:blk_1073741825_1001:java.io.IOException: Broken pipe
java.io.IOException: Broken pipe
	at sun.nio.ch.FileDispatcherImpl.write0(Native Method)
	at sun.nio.ch.SocketDispatcher.write(SocketDispatcher.java:47)
	at sun.nio.ch.IOUtil.writeFromNativeBuffer(IOUtil.java:93)
	at sun.nio.ch.IOUtil.write(IOUtil.java:65)
	at sun.nio.ch.SocketChannelImpl.write(SocketChannelImpl.java:470)
	at org.apache.hadoop.net.SocketOutputStream$Writer.performIO(SocketOutputStream.java:62)
	at org.apache.hadoop.net.SocketIOWithTimeout.doIO(SocketIOWithTimeout.java:141)
	at org.apache.hadoop.net.SocketOutputStream.write(SocketOutputStream.java:158)
	at org.apache.hadoop.net.SocketOutputStream.write(SocketOutputStream.java:116)
	at java.io.DataOutputStream.write(DataOutputStream.java:107)
	at org.apache.hadoop.thirdparty.protobuf.CodedOutputStream$OutputStreamEncoder.doFlush(CodedOutputStream.java:2930)
	at org.apache.hadoop.thirdparty.protobuf.CodedOutputStream$OutputStreamEncoder.flush(CodedOutputStream.java:2842)
	at org.apache.hadoop.thirdparty.protobuf.AbstractMessageLite.writeDelimitedTo(AbstractMessageLite.java:95)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.DataTransferSaslUtil.sendSaslMessage(DataTransferSaslUtil.java:580)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.DataTransferSaslUtil.sendSaslMessage(DataTransferSaslUtil.java:556)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.DataTransferSaslUtil.sendSaslMessage(DataTransferSaslUtil.java:549)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.DataTransferSaslUtil.sendGenericSaslErrorMessage(DataTransferSaslUtil.java:412)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslDataTransferClient.doSaslHandshake(SaslDataTransferClient.java:606)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslDataTransferClient.getSaslStreams(SaslDataTransferClient.java:446)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslDataTransferClient.send(SaslDataTransferClient.java:289)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslDataTransferClient.checkTrustAndSend(SaslDataTransferClient.java:236)
	at org.apache.hadoop.hdfs.protocol.datatransfer.sasl.SaslDataTransferClient.peerSend(SaslDataTransferClient.java:170)
	at org.apache.hadoop.hdfs.DFSUtilClient.peerFromSocketAndKey(DFSUtilClient.java:822)
	at org.apache.hadoop.hdfs.DFSClient.newConnectedPeer(DFSClient.java:3052)
	at org.apache.hadoop.hdfs.client.impl.BlockReaderFactory.nextTcpPeer(BlockReaderFactory.java:829)
	at org.apache.hadoop.hdfs.client.impl.BlockReaderFactory.getRemoteBlockReaderFromTcp(BlockReaderFactory.java:754)
	at org.apache.hadoop.hdfs.client.impl.BlockReaderFactory.build(BlockReaderFactory.java:381)
	at org.apache.hadoop.hdfs.DFSInputStream.getBlockReader(DFSInputStream.java:755)
	at org.apache.hadoop.hdfs.DFSInputStream.actualGetFromOneDataNode(DFSInputStream.java:1199)
	at org.apache.hadoop.hdfs.DFSInputStream$2.call(DFSInputStream.java:1171)
	at org.apache.hadoop.hdfs.DFSInputStream$2.call(DFSInputStream.java:1167)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Reply:
I could reproduce the test failure by closing `out` before sending error.
```
    } catch (IOException ioe) {
      out.close();
      sendGenericSaslErrorMessage(out, ioe.getMessage());
      throw ioe;
    }
```
Throwing the original exception as follows fixed the test case. Therefore your fix is really helpful.
```
    } catch (IOException ioe) {
      out.close();
      try {
        sendGenericSaslErrorMessage(out, ioe.getMessage());
      } catch (Exception e) {
        ioe.addSuppressed(e);
      }
      throw ioe;
    }
```

Reply:
Given this fix is commented in detail and creating the unit test seems costly (mock can do this, but the maintenance cost seems too much), I think we don't need a unit test to cover this case. Thanks.

Reply:
Thank you for confirming :bowing_man: